### ISSUE TITLE
Envoy loses connection to Pilot and does not reconnect

### ISSUE THREAD

**Issue by mandarjog** [istio/istio#7296]
This was observed in a build from 2018-7-16, after it was running for 4 days.

1. Deployment stopped working with 503s at ingress.
2. Ingress was unable to connect to Mixer, it was using old mixer address and disallowing access.
3. None of the other sidecars receiving any updates from pilot.
4. Pilot logs show that 
```
2018-07-20T19:11:58.550945Z     warn    ads     ADS: RDS: Send failure for &{Kubernetes sidecar 10.44.1.42 reviews-v1-7ff48f6566-t5j99.default default.svc.cluster.local 
map[POD_NAME:reviews-v1-7ff48f6566-t5j99 INTERCEPTION_MODE:REDIRECT istio:sidecar ISTIO_PROXY_VERSION:1.0.0 ISTIO_PROXY_SHA:istio-proxy:817e3f93a1e4d97788fc838910fb2f9e0503650b 
ISTIO_VERSION:release-1.0-20180716-16-37]}, closing grpc rpc error: code = Canceled desc = context canceled
```
indicating that reviews has disconnected, however reviews proxy does not see the disconnect.
Following is the output of netstat.
`reviewsV1 = 10.44.1.42` and `pilot =  10.44.0.17`
At pilot
```
root@istio-pilot-65dd6fb899-qkmqk:/# netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 127.0.0.1:9876          0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:15003           0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:15005           0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:15007           0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:15011           0.0.0.0:*               LISTEN
tcp        0      0 10.44.0.17:55862        10.47.244.7:15004       ESTABLISHED
tcp        0      0 10.44.0.17:15011        10.44.2.19:40414        ESTABLISHED
tcp        0      0 127.0.0.1:44956         127.0.0.1:15010         ESTABLISHED
tcp        0      0 10.44.0.17:55864        10.47.244.7:15004       ESTABLISHED
tcp        0      0 10.44.0.17:55858        10.47.244.7:15004       ESTABLISHED
tcp        0      0 10.44.0.17:15011        10.44.1.44:35548        ESTABLISHED
tcp        0      0 127.0.0.1:44958         127.0.0.1:15010         ESTABLISHED
tcp        0      0 10.44.0.17:55860        10.47.244.7:15004       ESTABLISHED
tcp        0      0 127.0.0.1:44962         127.0.0.1:15010         ESTABLISHED
tcp        0      0 10.44.0.17:44386        10.47.240.1:443         ESTABLISHED
tcp        0      0 127.0.0.1:44964         127.0.0.1:15010         ESTABLISHED
```

And at ** Reviews proxy **
```
istio-proxy@reviews-v1-7ff48f6566-t5j99:/$ netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp        0      0 127.0.0.1:15000         0.0.0.0:*               LISTEN
tcp        0      0 0.0.0.0:15001           0.0.0.0:*               LISTEN
tcp        0      0 10.44.1.42:35372        10.47.252.178:15011     ESTABLISHED
tcp6       0      0 :::9080                 :::*                    LISTEN
tcp6       0      0 127.0.0.1:35261         :::*                    LISTEN
udp        0      0 10.44.1.42:37892        10.47.240.114:9125      ESTABLISHED
udp        0      0 10.44.1.42:40804        10.47.240.114:9125      ESTABLISHED
udp        0      0 10.44.1.42:43076        10.47.240.114:9125      ESTABLISHED
udp        0      0 10.44.1.42:43904        10.47.240.114:9125      ESTABLISHED
udp        0      0 10.44.1.42:45297        10.47.240.114:9125      ESTABLISHED
Active UNIX domain sockets (servers and established)
Proto RefCnt Flags       Type       State         I-Node   Path
unix  2      [ ]         DGRAM                    8941069  @envoy_domain_socket_0@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
unix  2      [ ]         STREAM     CONNECTED     8939466
unix  2      [ ]         STREAM     CONNECTED     8940410
```

Pods ips
```
mjog-macbookpro:quotaExamples mjog$ k g pods -o wide
kubectl get pods -o wide
NAME                              READY     STATUS    RESTARTS   AGE       IP           NODE
details-v1-74b7764547-wjmwk       2/2       Running   0          4d        10.44.2.14   gke-cl1100-default-pool-bc10ad69-qs3k
productpage-v1-869dc5df57-86npk   2/2       Running   0          2h        10.44.1.44   gke-cl1100-default-pool-bc10ad69-4nb3
ratings-v1-8477d46bb7-fc5kg       2/2       Running   0          4d        10.44.0.13   gke-cl1100-default-pool-bc10ad69-hgpc
reviews-v1-7ff48f6566-t5j99       2/2       Running   0          2d        10.44.1.42   gke-cl1100-default-pool-bc10ad69-4nb3
reviews-v2-854cb576c4-5rn94       2/2       Running   0          4d        10.44.0.14   gke-cl1100-default-pool-bc10ad69-hgpc
reviews-v3-645498c858-sg9ln       2/2       Running   0          4d        10.44.2.15   gke-cl1100-default-pool-bc10ad69-qs3k
sleep-86f6b99f94-8lmxp            2/2       Running   0          4d        10.44.1.38   gke-cl1100-default-pool-bc10ad69-4nb3
mjog-macbookpro:quotaExamples mjog$ ik g pods -o wide
kubectl --namespace istio-system get pods -o wide
NAME                                       READY     STATUS    RESTARTS   AGE       IP           NODE
istio-citadel-64c559d5cd-mtxfb             1/1       Running   0          4d        10.44.1.32   gke-cl1100-default-pool-bc10ad69-4nb3
istio-egressgateway-6c8bd65bd9-qfjjk       1/1       Running   0          4d        10.44.1.31   gke-cl1100-default-pool-bc10ad69-4nb3
istio-galley-5b6556fb89-zwxnr              1/1       Running   0          4d        10.44.1.34   gke-cl1100-default-pool-bc10ad69-4nb3
istio-ingressgateway-b765dc74c-mpfj9       1/1       Running   0          22m       10.44.2.19   gke-cl1100-default-pool-bc10ad69-qs3k
istio-pilot-65dd6fb899-qkmqk               2/2       Running   0          2d        10.44.0.17   gke-cl1100-default-pool-bc10ad69-hgpc
istio-policy-7444897d44-nfmvv              2/2       Running   0          6h        10.44.0.19   gke-cl1100-default-pool-bc10ad69-hgpc
istio-sidecar-injector-574997d6c7-2vp4r    1/1       Running   0          4d        10.44.2.13   gke-cl1100-default-pool-bc10ad69-qs3k
istio-statsd-prom-bridge-59b45fd6d-qzxp2   1/1       Running   0          4d        10.44.1.30   gke-cl1100-default-pool-bc10ad69-4nb3
istio-telemetry-59484f4d98-s8zvg           2/2       Running   0          4d        10.44.1.28   gke-cl1100-default-pool-bc10ad69-4nb3
prometheus-ffd95f9f6-wkf4g                 1/1       Running   0          4d        10.44.0.12   gke-cl1100-default-pool-bc10ad69-hgpc
```

**Comment by rshriram** [comment]
is this error happening at ADS layer? Seems so. We had some changes recently in that area (specifically around fixing the race conditions). I don't see how that can help, but worth upgrading to latest release and checking it out.

**Comment by mandarjog** [comment]
The grpc stream is in half close mode, keep alive helps.

**Comment by mandarjog** [comment]
resolved by #7332
